src/gpu/d3d11/SDL_gpu_d3d11.c

```diff
@@ -3106,11 +3118,52 @@ static void D3D11_GetTransferData(
 	}
 	else /* TEXTURE */
 	{
-		SDL_memcpy(
-			((Uint8*) data) + copyParams->dstOffset,
-			(Uint8*) buffer->textureTransfer.data + copyParams->srcOffset,
-			copyParams->size
-		);
+        SDL_LockMutex(renderer->contextLock);
+
+        /* Read from the staging texture */
+        res = ID3D11DeviceContext_Map(
+            renderer->immediateContext,
+            buffer->textureTransfer.downloadTexture,
+            0,
+            D3D11_MAP_READ,
+            0,
+            &subresource
+        );
+        ERROR_CHECK_RETURN("Could not map texture for reading",)
+
+        dataPtr = (Uint8*) data + copyParams->dstOffset;
+
+        if (buffer->textureTransfer.downloadTightlyPacked)
+        {
+            SDL_memcpy(
+                dataPtr,
+                (Uint8*) subresource.pData + copyParams->srcOffset,
+                buffer->textureTransfer.downloadDepth * buffer->textureTransfer.downloadHeight * buffer->textureTransfer.downloadBytesPerRow
+            );
+        }
+        else
+        {
+            for (Uint32 depth = 0; depth < buffer->textureTransfer.downloadDepth; depth += 1)
+            {
+                for (Uint32 row = 0; row < buffer->textureTransfer.downloadHeight; row += 1)
+                {
+                    SDL_memcpy(
+                        dataPtr,
+                        (Uint8*) subresource.pData + copyParams->srcOffset + (depth * buffer->textureTransfer.downloadBytesPerDepthSlice) + (row * buffer->textureTransfer.downloadBytesPerRow),
+                        buffer->textureTransfer.downloadBytesPerRow
```

Comment:
I think to work as people expect the number of bytes copied here might need to be derived from the width of the resource instead of the stride. But I don't know how you would actually implement that, or whether you should.

Reply:
The download values are provided by the client in the DownloadFromTexture call and cached into these variables.